docs/design/panic-error.md

```diff
@@ -0,0 +1,231 @@
+errors are returned. This doesn't necessarily mean that panicking is right in
+this situation; it only means that returning errors comes with a cost that must
+be weighed against their benefits and the costs of panicking.
+
+## Reliability
+
+Panics are meant to be rare, exceptional occurrences that force an immediate,
+and potentially unexpected halt to all the goroutines running in the process.
+Libraries panicking are especially hazardous, because the context the panics
+occur in can be more expensive than the library anticipated. A library that
+panics too frequently or during regular operation is considered less reliable
+than a library that panics rarely and only in exceptional cases. When choosing
+between panics and errors for a specific exception, we should keep in mind how
+frequently that exception is expected to be encountered.
+
+Panics are so expensive because they are so difficult for the caller to
+effectively handle. While callers can use `recover` to handle panics, this is
+non-obvious enough and far enough from standard Go practice that it doesn't
+really help that much. Errors, on the other hand, have a rich and robust
+ecosystem of packages, patterns, and best practices for detecting and handling
+them. Using errors allows the caller to automatically fix or work around the
+issue: retrying transient errors, creating files that don't exist, or resolving
+other exceptions that occurred because of the environment the code was run in.
+These automated fixes or workarounds are essential to the resiliency and
+reliability of programs, and panics undercut the ability to take advantage of
+them.
+
+## Conspicuousness
+
+When returning an error as the only return value of a function, Go will not
+complain if the caller ignores the returned error:
+
+```go
+func myFunc() error {
+	return errors.New("this is why we can't have nice things")
+}
+
+func caller() {
+	// note we're not checking or handling the error here
+	// and the Go compiler doesn't tell us about the unused return value
+	myFunc()
+}
+```
+
+This is most famously seen with the
+[helper/schema.ResourceData.Set][resourcedata-set] function in v2 of the
+Terraform Plugin SDK. This function sets a takes a path and a value, and then
+sets that value in state at that path. It returns an error if the path is
+invalid, if the value is an invalid type, or if the type of the value doesn't
+match the type expected at that path. This function is ubiquitous in provider
+development. Unfortunately, possibly because it's so ubiquitous, provider
+developers often forget to check the error value returned, leading to the
+exception not being surfaced or addressed, and unexpected behavior from the
+provider.
+
+While it's easy to assert that provider developers should just check the error,
+the API should be designed such that it's hard to use incorrectly, and
+incorrect usage should be loud and yield noticeable feedback.
```

Comment:
We can also strongly recommend (and include in terraform-provider-scaffolding) the use of a linter to ensure errors are always checked. It's not the perfect solution, but it's another option.

Reply:
Think I addressed this in the text.